{"data": {"repository": {"pullRequest": {"id": "MDExOlB1bGxSZXF1ZXN0NDY2NDIxOTY2", "number": 2702, "reviewThreads": {"totalCount": 2, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wOC0xMlQwMDozMDowMVrOEXbXXw==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wOC0xMlQwMDo1MDoyNVrOEXbmBQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjkzMDAxMDU1OnYy", "diffSide": "RIGHT", "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "isResolved": false, "comments": {"totalCount": 2, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wOC0xMlQwMDozMDowMVrOG_NskQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wOC0xMlQxNzoyMDo1MFrOG_rEOw==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODkzNzg3Mw==", "bodyText": "Should the exception still be thrown after the handler is called ?", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r468937873", "createdAt": "2020-08-12T00:30:01Z", "author": {"login": "Maithem"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();\n                         throw we;", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 5}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2OTQxOTA2Nw==", "bodyText": "the reason for throwing the exception is some shutdown handlers may just be no-ops (by intent), in which case we want to still fail the operation.", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r469419067", "createdAt": "2020-08-12T17:20:50Z", "author": {"login": "hisundar"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();\n                         throw we;", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODkzNzg3Mw=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 5}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjkzMDA0ODA1OnYy", "diffSide": "RIGHT", "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "isResolved": false, "comments": {"totalCount": 13, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wOC0xMlQwMDo1MDoyNVrOG_OB3Q==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wOC0xN1QyMzo0NTowMVrOHB-dEQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ==", "bodyText": "Should we change the layoutHelper instead of here? Here it's fetchLayout, but other operations can still hit WCE", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r468943325", "createdAt": "2020-08-12T00:50:25Z", "author": {"login": "WenbinZhu"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2OTAxNTUyNQ==", "bodyText": "+1", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r469015525", "createdAt": "2020-08-12T05:36:59Z", "author": {"login": "annym"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2OTQxOTMwOA==", "bodyText": "will do. thanks", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r469419308", "createdAt": "2020-08-12T17:21:16Z", "author": {"login": "hisundar"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MDE1Njk2NA==", "bodyText": "done. thanks", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r470156964", "createdAt": "2020-08-13T18:21:30Z", "author": {"login": "hisundar"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MDE2MjUyNA==", "bodyText": "Looks like sometimes this method, fetchLayout, is called inside LayoutHelper, so we will invoke SystemDownHandler twice right? Will this be a problem?", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r470162524", "createdAt": "2020-08-13T18:31:53Z", "author": {"login": "WenbinZhu"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MDM2NDc4Ng==", "bodyText": "systemDownHandler are supposed to exit the JVM, so not unless there is another bug right?", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r470364786", "createdAt": "2020-08-14T01:40:19Z", "author": {"login": "hisundar"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MDM2OTEzMQ==", "bodyText": "@hisundar Actually I feel this catch block will never be executed. The exceptions should come from layoutFuture.get(), but that will be wrapped in ExecutionException, which we check in the next catch block, right?", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r470369131", "createdAt": "2020-08-14T01:59:08Z", "author": {"login": "WenbinZhu"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MTcyOTY2NQ==", "bodyText": "@annym @WenbinZhu Since Sundar is out, I fixed it in two places where the issue might happen.\nOne way to get WCE is to invalidateLayout() after our local current node has joined another cluster and it had an identity (clusterId) before that. That WCE is hit when we call invalidateLayout() inside a layoutHelper. checkClusterId inside fetchLayout will throw WrongClusterException that will not be handled, so the runtime layout future will complete with ExecutionException with the cause of the WrongClusterException. Then when in layoutHelper we call getLayoutUninterruptibly() and we call get on the this layout future, we will hit ExecutionException with the the cause of WrongClusterException. Here we should call a systemDownHandler and then propagate this exception (since downhandler might be a no-op).\nAnother way to get WCE is in function.apply in the layoutHelper. This can happen if the function is an RPC and the client message was stamped with one clusterId but the server router on the remote node has another clusterId. Here we should also call a systemDownHandler and propagate the exception.", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r471729665", "createdAt": "2020-08-17T19:32:11Z", "author": {"login": "PavelZaytsev"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MTc0ODc0MA==", "bodyText": "@PavelZaytsev invalidateLayout() can be called outside of AbstractView, which means if users use runtime.invalidateLayout() directly, then SystemDownHandler will not be trigger, because you don't trigger it in fetchLayout(). So I think the  invocation of SystemDownHandler should be put into CorfuRuntime::fetchLayout instead of AbstractView::getLayoutUninterruptibly(), and also we already have SystemDownHandler invocation logic in fetchLayout()", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r471748740", "createdAt": "2020-08-17T20:11:42Z", "author": {"login": "WenbinZhu"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MTc2MjY1MQ==", "bodyText": "@WenbinZhu. Ok, I will call it in fetchLayout then every client would have to call system down handler. We still need to call it in the layoutHelper for the RPC case though.  I removed it from getLayoutUninterruptibly, so it will just get propagated as ExecutionException with a RuntimeException cause after the systemDownHandler is called.", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r471762651", "createdAt": "2020-08-17T20:39:58Z", "author": {"login": "PavelZaytsev"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MTc3NzEyMg==", "bodyText": "@PavelZaytsev Thanks, but I think there is another problem as I mentioned in the initial commit.  Looks like catch (WrongClusterException we) will never be reached (a bug in the existing code), because execption is wrapped in the ExecutionException. I think this exception handling block should be moved to the catch (ExecutionException ee)  block below, otherwise it will not work.", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r471777122", "createdAt": "2020-08-17T21:10:46Z", "author": {"login": "WenbinZhu"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MTc5MjgyNQ==", "bodyText": "@WenbinZhu catch (WrongClusterException we) is actually getting caught right now and is getting propagated because it is coming from checkClusterId, so it is not wrapped into ExecutionException. You are probably referring to layoutFuture.get() that is being called after getLayout(). But this one can not fail with a WrongClusterException because it is a LAYOUT_REQUEST, and we ignore a clusterId check for it.\nLAYOUT_REQUEST(10, new TypeToken<CorfuPayloadMsg<Long>>(){}, true, true)", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r471792825", "createdAt": "2020-08-17T21:46:24Z", "author": {"login": "PavelZaytsev"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3MTgzMzg3Mw==", "bodyText": "I see", "url": "https://github.com/CorfuDB/CorfuDB/pull/2702#discussion_r471833873", "createdAt": "2020-08-17T23:45:01Z", "author": {"login": "WenbinZhu"}, "path": "runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java", "diffHunk": "@@ -1140,6 +1140,7 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {\n                     } catch (WrongClusterException we) {\n                         // It is futile trying to re-connect to the wrong cluster\n                         log.warn(\"Giving up since cluster is incorrect or reconfigured!\");\n+                        parameters.getSystemDownHandler().run();", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ2ODk0MzMyNQ=="}, "originalCommit": {"oid": "d0c656f7f33822f651a4d7b64bfc9681218b6c45"}, "originalPosition": 4}]}}]}}}, "rateLimit": {"limit": 5000, "remaining": 1919, "cost": 1, "resetAt": "2021-11-12T11:57:46Z"}}}